Implement Zobrist hashing (closes #55, fixes #3)

niklasf opened the pull request:
The previous position hasher was *not* accounting for

* **en-passant**,
* **checks given in three-check** and
* **crazyhouse data**.

It was therefore incorrectly giving some threefold repetitions.
This is now fixed.

In order to account for **castling**, the previous implementation was
clearing the transposition table on castling moves. Since the
transposition table is also used for the 50-move rule, this incorrectly
reset the half move counter (issue #3). This is now fixed.

In order to update the hash after a move (with castling rights and
en-passant privileges), **the hasher now needs to work on a fully
finialized situation**.

Zobrist hashing is based on a set of **pre-agreed random bitmasks**. The
particular ones chosen are **compatible with the Polyglot opening book
format**. This allows testing against given reference values. (Additionally
Polyglot opening book reading would now be easy to implement, but we
don't really have a use case for this, at the moment.)

---

Please review! Comments and suggestions welcome.

ornicar commented:
Please merge ornicar/master back so we get travis CI on this.

ornicar commented on src/main/scala/Hash.scala:
```diff
@@ -1,19 +1,643 @@
+class Hash(
+    size: Int) {
 
-object Hash {
```
`object` is right here, `class` isn't. Think of `object` as a singleton.

ornicar commented on src/main/scala/Hash.scala:
```diff
@@ -1,19 +1,643 @@
+
+  private def crazyPocketIndex(role: Role, color: Color, count: Int) = {
+    assert(0 < count && count <= 16)
+    assert(roleIndex(role) < 5)
```
you won't find a single `assert` in scalachess or lila, because I don't like things that blow up at runtime.

Instead, you can express the uncompleteness of the function by returning an `Option[Int]`.

ornicar commented on src/main/scala/Hash.scala:
```diff
@@ -1,19 +1,643 @@
+  private[chess] def hexToBytes(str: String): PositionHash = {
+    str.grouped(2).map(cc =>
+      (Character.digit(cc(0), 16) << 4 | Character.digit(cc(1), 16)).toByte
+    ).toArray take size
```
maybe `take(size)` and `toArray` second. All that code will run 600 times by per second, probably 1000 in a few months.

ornicar replied:
actually I checked and this function only runs at startup.

ornicar commented on src/main/scala/Hash.scala:
```diff
@@ -1,19 +1,643 @@
+
+    // According to FIDE rules en-passant needs to be completely legal in order
+    // to make a difference with regards to threefold repetition.
+    val ep = situation.moves.values.flatten.find(_.enpassant) match {
```
this might constitute a performance regression, unless another part of the code was already requesting move generation for this position.

niklasf replied:
I wonder if we could at least reuse the moves generated here, instead of doing-it on a throw-away Situation.

ornicar commented:
That's great stuff, hurray for correctness; and for tests. I'm concerned about performance implications since that stuff runs on every move. Will write some benchmarks.

ornicar commented:
chess.PlayPerfTest determines how many games can be played per second on a single thread.

Current master: 245 games per second
This branch: 170 games per second

That's a 30% performance drop overall.

ornicar commented on src/main/scala/Hash.scala:
```diff
@@ -1,19 +1,643 @@
 
+  // The first 8 bytes of the following masks are compatible with the Polyglot
+  // opening book format. The second half is generated randomly.
+
```
do we really need 16 bytes per mask here? The runtime is only using 3 bytes anyway (to reduce DB pressure)

niklasf replied:
Indeed we don't. And in fact all the literals are parsed by hexToBytes at startup and all but 3 bytes per hash are dropped.

The thing I have in mind is the Opening explorer (ornicar/lila#1528). It could now simply instanciate `new Hash(/*size*/ 16)` and produce globally unique hashes for hundreds of millions of games.

ornicar replied:
ok

niklasf commented:
Thanks for the review! I pushed an improved (and rebased) version and addressed the remaining comments.

Performance regression (especially because of the move generation for en-passant) are still a thing, though. Trying to figure something out ...

niklasf commented:
After the discussed optimization, performance drop on my machine is: 376 games per second before, 323 after, -14%. 338 without en-passant.

Looks like there is still potential for improvement.

niklasf commented:
Optimized it some more. We are now in the -10% range. Also lists seam to be slightly better suited to store the selected masks than arrays.
